src/QirRuntime/lib/QIR/callables.cpp

```diff
@@ -11,47 +11,36 @@
 #include "qirTypes.hpp"
 #include "quantum__rt.hpp"
 
+/*==============================================================================
+    Implementation of quantum__rt__tuple_* and quantum__rt__callable_*
+==============================================================================*/
 extern "C"
 {
-    QirTupleHeader* quantum__rt__tuple_create(int64_t size)
+    PTuple quantum__rt__tuple_create(int64_t size)
     {
-        assert(size >= sizeof(QirTupleHeader));
-        char* buffer = new char[size];
-
-        // at the beginning of the buffer place QirTupleHeader
-        QirTupleHeader* th = reinterpret_cast<QirTupleHeader*>(buffer);
-        th->refCount = 1;
-        th->tupleSize = size;
-
-        return th;
+        assert(size >= 0 && size < std::numeric_limits<int>::max());
+        return QirTupleHeader::Create(static_cast<int>(size))->AsTuple();
     }
 
-    void quantum__rt__tuple_reference(QirTupleHeader* th)
+    void quantum__rt__tuple_reference(PTuple tuple)
     {
-        if (th == nullptr)
+        if (tuple == nullptr)
         {
             return;
         }
-        assert(th->refCount > 0); // no resurrection of deleted tuples
-        ++th->refCount;
+        QirTupleHeader* th = QirTupleHeader::GetHeader(tuple);
+        (void)th->AddRef();
     }
 
-    void quantum__rt__tuple_unreference(QirTupleHeader* th)
+    void quantum__rt__tuple_unreference(PTuple tuple)
     {
-        if (th == nullptr)
+        if (tuple == nullptr)
         {
             return;
         }
 
-        const long ref = --th->refCount;
-        assert(ref >= 0);
-
-        if (ref == 0)
-        {
-            char* buffer = reinterpret_cast<char*>(th);
-            delete[] buffer;
-        }
-        th = nullptr;
+        QirTupleHeader* th = QirTupleHeader::GetHeader(tuple);
+        (void)th->Release();

```

Comment:
You do not want to assert that refcount is >=0?

Reply:
If refcount dropped to zero before, the memory would be released and potentially already reclaimed. The assert would be checking a random value. The similar assert in AddRef has same problem but I thought I'd have it to document that objects aren't expected to be resurrected.